### PR TITLE
feat: Localization module

### PR DESCRIPTION
**Pull request by ahmed3mar**
Closes goravel/goravel#39

## 📑 Description
Implement translations

Related PR
https://github.com/goravel/goravel/pull/195

First we can need to set config in app ```"locale": "en"``` and ```"fallback_locale": "en",```
then register the provider ```&translation.ServiceProvider{},```

we can load translations for all languages at once by using ```facades.Translation().Load("resources/lang")```
or for specific language ```facades.Translation("ar").Load("resources/lang/ar")```

there are several ways to use translations,

First in request context for example
```go
func (r *UserController) Show(ctx http.Context) {
	ctx.Response().Success().Json(http.Json{
		"Hello": "Goravel",
		"errors": ctx.Trans("validation.required", http.Json{
			"attribute": "name",
		}, "ar"),
	})
}
```

the first parameter is required but replace and language is optional

we can set or get current language by using
```ctx.SetLocale("en")``` or ```ctx.GetLocale()```

second from facades
```go
func (r *UserController) Show(ctx http.Context) {
	ctx.Response().Success().Json(http.Json{
		"Hello": "Goravel",
		"translate": facades.Translation().Get("validation.required", "en", http.Json{
			"attribute": "name",
		}),
	})
}
```
first and second parameters are required

or 
```go
facades.Language("ar").Get("validation.required", http.Json{
	"attribute": "name",
})
```
to get translations for specific language

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

File: contracts/translation/translation.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	//
+	Language(locale string) Language
+	// Add translations to application.
+	Add(name, locale string, val any)
+	// Get translation from application.
+	Get(path string, locale string, options ...any) string
+}
+
+//go:generate mockery --name=Language
+type Language interface {
```
Comment (hwbrzzl) [inline contracts/translation/translation.go:17]: Q: Is there a similar feature in Laravel? I can't seem to find it.

File: facades/translation.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package facades
+
+import "github.com/goravel/framework/contracts/translation"
+
+func Translation() translation.Translation {
```
Comment (hwbrzzl) [inline facades/translation.go:5]: S: It's `Lang` in Laravel: https://github.com/laravel/framework/blob/c262c1ffce93c5fa4b9e00da1a54d8c02ddddcf6/src/Illuminate/Support/Facades/Lang.php#L44, and it seems that this method needs to add a `context.Context` parameter.

File: translation/utils.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	// Iterate over the files and filter out directories.
+	for _, file := range files {
+		if !file.IsDir() {
+			continue
+		}
+
+		// Add the directory to the list of directories.
+		dirs = append(dirs, filepath.Join(root, file.Name()))
```
Comment (hwbrzzl) [inline translation/utils.go:53]: Q: What happens if `file` has subdirectories?

File: translation/language.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+var _ translation.Language = &Language{}
+
+type Language struct {
+	vip *viper.Viper
```
Comment (hwbrzzl) [inline translation/language.go:18]: Q: Is it necessary to use `viper` here?

File: contracts/translation/translation.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package translation
+
+//go:generate mockery --name=Translation
+type Translation interface {
+	// Load translations from path.
+	Load(path string, locale ...string) error
+	GetDefaultLocale() string
```
Comment (hwbrzzl) [inline contracts/translation/translation.go:7]: S: consistent with Laravel: https://github.com/laravel/framework/blob/c262c1ffce93c5fa4b9e00da1a54d8c02ddddcf6/src/Illuminate/Translation/Translator.php#L404
```suggestion
	Locale() string
```

File: contracts/translation/translation.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	//
+	Language(locale string) Language
+	// Add translations to application.
+	Add(name, locale string, val any)
```
Comment (hwbrzzl) [inline contracts/translation/translation.go:11]: S: Consistent with Laravel: https://github.com/laravel/framework/blob/aef89589ea70e0081c139b06550220cc75f20ea6/src/Illuminate/Translation/Translator.php#L256
```suggestion
	AddLines(lines map[string]any, locale string)
```

File: contracts/translation/translation.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	// Get translation from application.
+	Get(path string, locale string, options ...any) string
```
Comment (hwbrzzl) [inline contracts/translation/translation.go:13]: S: We can add a `Option` struct, just like: [validation/options.go](https://github.com/goravel/framework/blob/master/validation/options.go)
```suggestion
	Get(path string, options ...Option) string
```

**Review by hwbrzzl — COMMENTED** [review]
A great feature, we can associate this PR to https://github.com/goravel/goravel/issues/39.

File: translation/language.go (hunk 4 above)
Comment (ahmed3mar) [reply translation/language.go:19]: Yup, because we don't have Arr::set yet

**Comment by hwbrzzl** [comment]
Hi @ahmed3mar Are you still dealing with this PR, please?

**Comment by hwbrzzl** [comment]
Closed this PR first, feel free to reopen it if needed.

**Pull request closed without merging**